File: src/main/scala/io/izzel/taboolib/util/item/inventory/ClickEvent.java
package io.izzel.taboolib.util.item.inventory;

import com.google.common.collect.Lists;
import io.izzel.taboolib.util.lite.Servers;
import org.bukkit.entity.Player;
import org.bukkit.event.Cancellable;
import org.bukkit.event.Event;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryDragEvent;
import org.bukkit.event.inventory.InventoryEvent;
import org.bukkit.event.inventory.InventoryInteractEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.function.Consumer;

/**
 * 界面构建工具 {@link MenuBuilder} 中的点击事件
 *
 * @Author 坏黑
 * @Since 2019-05-21 18:09
 */
public class ClickEvent {

    private final ClickType clickType;
    private final Event event;
    private final char slot;

    public ClickEvent(ClickType clickType, Event event, char slot) {
        this.clickType = clickType;
        this.event = event;
        this.slot = slot;
    }

    @NotNull
    public ClickEvent onClick(Consumer<InventoryClickEvent> consumer) {
        consumer.accept(castClick());
        return this;
    }

    @NotNull
    public ClickEvent onDrag(Consumer<InventoryDragEvent> consumer) {
        consumer.accept(castDrag());
        return this;
    }

    /**
     * 获取受影响的物品
     */
    @NotNull
    public List<ItemStack> getAffectItems() {
        return clickType == ClickType.CLICK ? Servers.getAffectItemInClickEvent((InventoryClickEvent) event) : Lists.newArrayList();
    }

    /**
     * 转换为点击事件
     */
    @NotNull
    public InventoryClickEvent castClick() {
        return (InventoryClickEvent) event;
    }

    /**
     * 转换为拖动事件
     */
    @NotNull
    public InventoryDragEvent castDrag() {
        return (InventoryDragEvent) event;
    }

    /**
     * 获取点击位置
     */
    public int getRawSlot() {
        return clickType == ClickType.CLICK ? castClick().getRawSlot() : -1;
    }

    /**
     * 获取点击位置（文本映射类型）
     */
    public char getSlot() {
        return slot;
    }

    /**
     * 获取点击方式
     */
    public ClickType getClickType() {
        return clickType;
    }

    /**
     * 获取点击者（玩家）实例
     */
    public Player getClicker() {
        return (Player) ((InventoryInteractEvent) event).getWhoClicked();
    }

    /**
     * 获取背包实例
     */
    public Inventory getInventory() {
        return ((InventoryEvent) event).getInventory();
    }

    /**
     * 是否取消点击事件
     */
    public void setCancelled(boolean c) {
        ((Cancellable) event).setCancelled(true);
    }

    /**
     * 点击事件是否被取消
     */
    public boolean isCancelled() {
        return ((Cancellable) event).isCancelled();
    }

    /**
     * 获取点击物品（仅限 CLICK 点击方式）
     */
    @Nullable
    public ItemStack getCurrentItem() {
        return clickType == ClickType.CLICK ? castClick().getCurrentItem() : null;
    }

    /**
     * 设置点击物品（仅限 CLICK 点击方式）
     * @param item 物品实例
     */
    public void setCurrentItem(ItemStack item) {
        if (clickType == ClickType.CLICK) {
            castClick().setCurrentItem(item);
        }
    }
}

File: module/bukkit/bukkit-ui/src/main/kotlin/taboolib/module/ui/ClickEvent.kt
package taboolib.module.ui

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryDragEvent
import org.bukkit.event.inventory.InventoryInteractEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.InventoryView
import org.bukkit.inventory.ItemStack
import taboolib.common.util.t
import taboolib.module.ui.type.Chest
import taboolib.module.ui.virtual.RemoteInventory
import taboolib.module.ui.virtual.VirtualInventoryInteractEvent

/**
 * @author 坏黑
 * @since 2019-05-21 18:09
 */
class ClickEvent(private val bukkitEvent: InventoryInteractEvent, val clickType: ClickType, val slot: Char, val builder: Chest) {

    private val cancelCallbacks = mutableListOf<java.util.function.Consumer<Array<StackTraceElement>>>()

    val clicker: Player
        get() = bukkitEvent.whoClicked as Player

    val inventory: Inventory
        get() = bukkitEvent.inventory

    val view: InventoryView
        get() = bukkitEvent.view

    /** 影响物品 */
    val affectItems: List<ItemStack>
        get() = if (clickType === ClickType.CLICK) clickEvent().getAffectItems() else emptyList()

    /** 取消事件 */
    var isCancelled: Boolean
        get() = bukkitEvent.isCancelled
        set(value) {
            bukkitEvent.isCancelled = value
            if (cancelCallbacks.isNotEmpty()) {
                val stackTrace = Thread.currentThread().stackTrace
                cancelCallbacks.forEach { it.accept(stackTrace) }
            }
        }

    /**
     * 注册取消回调，当事件取消状态被设置时触发
     * @param callback 回调函数，参数为调用堆栈
     */
    fun onCancel(callback: java.util.function.Consumer<Array<StackTraceElement>>): ClickEvent {
        cancelCallbacks += callback
        return this
    }

    /** 点击位置 */
    val rawSlot: Int
        get() = when (clickType) {
            ClickType.CLICK -> clickEvent().rawSlot
            ClickType.VIRTUAL -> virtualEvent().clickSlot
            ClickType.DRAG -> {
                val rawSlots = dragEvent().rawSlots
                if (rawSlots.size == 1) rawSlots.first() else -1
            }
        }

    /** 键盘按键 */
    val hotbarKey: Int
        get() = when (clickType) {
            ClickType.CLICK -> clickEvent().hotbarButton
            ClickType.VIRTUAL -> virtualEvent().hotbarKey
            else -> -1
        }

    /** 获取或设置点击物品 */
    var currentItem: ItemStack?
        get() = when (clickType) {
            ClickType.CLICK -> clickEvent().currentItem
            ClickType.VIRTUAL -> virtualEvent().clickItem
            else -> null
        }
        set(item) {
            when (clickType) {
                ClickType.CLICK -> {
                    clickEvent().currentItem = item
                }
                ClickType.VIRTUAL -> {
                    inventory.setItem(virtualEvent().clickSlot, item)
                }
                else -> {}
            }
        }

    /** 获取或设置指针物品 */
    var cursorItem: ItemStack?
        get() = when (clickType) {
            ClickType.CLICK -> clickEvent().cursor
            ClickType.DRAG -> dragEvent().cursor
            else -> null
        }
        set(item) {
            when (clickType) {
                ClickType.CLICK -> {
                    clickEvent().whoClicked.setItemOnCursor(item)
                }
                ClickType.DRAG -> {
                    dragEvent().cursor = item
                }
                else -> {}
            }
        }

    /** 获取物品 */
    fun getItem(slot: Char): ItemStack? {
        val idx = builder.slots.flatten().indexOf(slot)
        return if (idx in 0 until inventory.size) inventory.getItem(idx) else null
    }

    /** 获取物品列表 */
    fun getItems(slot: Char): List<ItemStack> {
        return builder.slots.flatten().mapIndexedNotNull { index, c -> if (c == slot) inventory.getItem(index) ?: ItemStack(Material.AIR) else null }
    }

    /** 转换为点击事件 */
    fun clickEvent(): InventoryClickEvent {
        if (clickType != ClickType.CLICK) {
            error(
                """
                    clickEvent() 无法在 "$clickType" 动作中使用。
                    clickEvent() is not available in "$clickType" action.
                """.t()
            )
        }
        return bukkitEvent as InventoryClickEvent
    }

    /** 安全转换为点击事件 */
    fun clickEventOrNull(): InventoryClickEvent? {
        return bukkitEvent as? InventoryClickEvent
    }

    /** 转换为拖拽事件 */
    fun dragEvent(): InventoryDragEvent {
        if (clickType != ClickType.DRAG) {
            error(
                """
                    dragEvent() 无法在 "$clickType" 动作中使用。
                    dragEvent() is not available in "$clickType" action.
                """.t()
            )
        }
        return bukkitEvent as InventoryDragEvent
    }

    /** 安全转换为拖拽事件 */
    fun dragEventOrNull(): InventoryDragEvent? {
        return bukkitEvent as? InventoryDragEvent
    }

    /** 转换为虚拟点击事件 */
    fun virtualEvent(): RemoteInventory.ClickEvent {
        if (clickType != ClickType.VIRTUAL) {
            error(
                """
                    virtualEvent() 无法在 "$clickType" 动作中使用。
                    virtualEvent() is not available in "$clickType" action.
                """.t()
            )
        }
        return (bukkitEvent as VirtualInventoryInteractEvent).clickEvent
    }

    /** 安全转换为虚拟点击事件 */
    fun virtualEventOrNull(): RemoteInventory.ClickEvent? {
        return (bukkitEvent as? VirtualInventoryInteractEvent)?.clickEvent
    }

    /** 用安全的方式处理点击事件 */
    fun onClick(consumer: InventoryClickEvent.() -> Unit): ClickEvent {
        if (clickType == ClickType.CLICK) {
            consumer(clickEvent())
        }
        return this
    }

    /** 用安全的方式处理拖拽事件 */
    fun onDrag(consumer: InventoryDragEvent.() -> Unit): ClickEvent {
        if (clickType == ClickType.DRAG) {
            consumer(dragEvent())
        }
        return this
    }

    /** 用安全的方式处理虚拟点击事件 */
    fun onVirtualClick(consumer: RemoteInventory.ClickEvent.() -> Unit): ClickEvent {
        if (clickType == ClickType.VIRTUAL) {
            consumer(virtualEvent())
        }
        return this
    }
}